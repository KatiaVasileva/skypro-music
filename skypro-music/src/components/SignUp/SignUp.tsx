"use client";

import Image from "next/image";
import styles from "./SignUp.module.css";
import { useAppDispatch, useAppSelector } from "@/store/store";
import { ChangeEventHandler, MouseEventHandler, useState } from "react";
import { signup } from "@/store/features/userSlice";
import { useRouter } from "next/navigation";
import Link from "next/link";

export default function SignUp() {
  const dispatch = useAppDispatch();
  const [formData, setFormData] = useState({
    email: "",
    password: "",
    username: "",
    passwordRepeat: "",
  });

  const requestError = useAppSelector((state) => state.user.errorMessage);
  const [error, setError] = useState("");
  const errorMessage = error ? error : requestError;

  const handleOnChange: ChangeEventHandler<HTMLInputElement> = (event) => {
    setError("");
    const { name, value } = event.target;
    setFormData({ ...formData, [name]: value });
  };

  const handleSignUp: MouseEventHandler<HTMLButtonElement> = async (event) => {
    event.preventDefault();

    console.log(formData.password);
    console.log(formData.passwordRepeat);
    console.log(formData.password !== formData.passwordRepeat);

    if (!formData.username.trim()) {
      setError("Введите имя");
      return;
    }

    if (formData.username.length < 3) {
      setError("Имя пользователя должно быть более 3 символов");
      return;
    }

    if (!formData.email.trim()) {
      setError("Введите почту");
      return;
    }

    if (!formData.email.includes("@")) {
      setError("Неверный формат почты");
      return;
    }

    if (!formData.password.trim()) {
      setError("Введите пароль");
      return;
    }

    if (formData.password.length < 6) {
      setError("Пароль должен быть более 6 символов");
      return;
    }

    if (formData.password !== formData.passwordRepeat) {
      setError("Пароли не совпадают");
      return;
    }

    try {
      await dispatch(
        signup({
          email: formData.email,
          password: formData.password,
          username: formData.username,
        })
      ).unwrap();
    } catch (error) {
      if (error instanceof Error) {
        console.error(error.message);
      }
    }
  };

  return (
    <div className={styles.wrapper}>
      <div className={styles.container}>
        <div className={styles.block}>
          <form className={styles.formLogin}>
            <a href="../">
              <div className={styles.logo}>
                <Image
                  className={styles.logoImage}
                  src="/img/logo_modal.png"
                  alt="logo"
                  priority
                  width={113.33}
                  height={43}
                />
              </div>
            </a>
            <input
              className={styles.inputFirstChild}
              type="text"
              name="username"
              value={formData.username}
              placeholder="Имя пользователя"
              onChange={handleOnChange}
              autoComplete="username"
            />
            <input
              className={styles.input}
              type="text"
              name="email"
              value={formData.email}
              placeholder="Почта"
              onChange={handleOnChange}
              autoComplete="email"
            />
            <input
              className={styles.input}
              type="password"
              name="password"
              value={formData.password}
              onChange={handleOnChange}
              placeholder="Пароль"
              autoComplete="current-password"
            />
            <input
              className={styles.input}
              type="password"
              name="passwordRepeat"
              onChange={handleOnChange}
              placeholder="Повторите пароль"
              autoComplete="current-password"
            />
            {errorMessage && (
              <div className={styles.errorBox}>
                <p className={styles.error}>{errorMessage}</p>
              </div>
            )}
            <button className={styles.signupButton} onClick={handleSignUp}>
              <Link className={styles.signupButtonLink} href="/signin">
                Зарегистрироваться
              </Link>
            </button>
          </form>
        </div>
      </div>
    </div>
  );
}
