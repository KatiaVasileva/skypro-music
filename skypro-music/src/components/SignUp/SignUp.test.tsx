import ReduxProvider from "@/store/ReduxProvider";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom";
import SignUp from "./SignUp";

describe("Signup component", () => {
  it("renders correctly", () => {
    const { container } = render(
      <ReduxProvider>
        <SignUp />
      </ReduxProvider>
    );
    expect(container).toMatchSnapshot();
  });
});
