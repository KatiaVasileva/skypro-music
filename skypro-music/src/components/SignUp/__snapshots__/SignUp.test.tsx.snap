// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup component renders correctly 1`] = `
<div>
  <div
    class="wrapper"
  >
    <div
      class="container"
    >
      <div
        class="block"
      >
        <form
          class="formLogin"
        >
          <a
            href="../"
          >
            <div
              class="logo"
            >
              <img
                alt="logo"
                class="logoImage"
                data-nimg="1"
                decoding="async"
                fetchpriority="high"
                height="43"
                src="/_next/image?url=%2Fimg%2Flogo_modal.png&w=256&q=75"
                srcset="/_next/image?url=%2Fimg%2Flogo_modal.png&w=128&q=75 1x, /_next/image?url=%2Fimg%2Flogo_modal.png&w=256&q=75 2x"
                style="color: transparent;"
                width="113.33"
              />
            </div>
          </a>
          <input
            autocomplete="username"
            class="inputFirstChild"
            name="username"
            placeholder="Имя пользователя"
            type="text"
            value=""
          />
          <input
            autocomplete="email"
            class="input"
            name="email"
            placeholder="Почта"
            type="text"
            value=""
          />
          <input
            autocomplete="current-password"
            class="input"
            name="password"
            placeholder="Пароль"
            type="password"
            value=""
          />
          <input
            autocomplete="current-password"
            class="input"
            name="passwordRepeat"
            placeholder="Повторите пароль"
            type="password"
          />
          <button
            class="signupButton"
          >
            <a
              class="signupButtonLink"
              href="/signin"
            >
              Зарегистрироваться
            </a>
          </button>
        </form>
      </div>
    </div>
  </div>
</div>
`;
